docs/codeExecution.md

```diff
@@ -0,0 +1,89 @@
+
+### `container` execution mode
+
+The `container` execution mode is currently used when PrairieLearn is running in production. It uses Docker to provide a degree of isolation from both PrairieLearn and other courses.
+
+Instead of using a pool of zygotes as described above, it actually maintains a pool of Docker containers, each of which runs a simple Node script (the _executor_), which in turn runs a Python zygote. The Node script listens for requests from PrairieLearn and essentially just forwards them to the Python process. You may ask, "Why not just run the zygote as the primary process in the container?" Well, starting up a Docker container is significantly more expensive than starting up a Python interpreter. Given that we ocasionally want to completely restart the Python worker, such as when it encounters an error, having an additional level of indirection allows us to gracefully restart the Python process inside the Docker container without having to restart the entire Docker container.
+
+This mode also allows us to isolate one course from another so that course A cannot see content from course B, and vice versa. To achieve this, we take advantage of bind mounts. When creating a container, PrairieLearn also creates a special directory on the host, and then mounts that directory to `/course` in the container. To execute content for course A, PrairieLearn first bind mounts that course's directory to the container's host directory. This is transitive to the container, which will now see that course's content at `/course`. In the future, when a different course's code needs to be executed in that container, PrairieLearn will simply update the bind mount to point to the other course.
+
+## Code execution in practice
+
+So far, this discussion has been pretty abstract. But what about all the actual code that underpins this stuff? Fear not, dear reader, we haven't forgotten about that!
+
+### Code callers
+
+A _code caller_ serves as an abstraction on top of the different execution modes above and hides the implementation details of exactly how code is executed. There are currently two different types of code callers, referred to here by the filenames of their implementations.
+
+- `lib/code-caller-docker` handles executing code inside of Docker containers, as required by the `container` execution mode.
+- `lib/code-caller-python` handles executing Python processes directly, as required by the `native` execution mode.
+
+The primary external interface of these callers is the `call()` function, which takes five arguments:
+
+- `type`: the type of code being executed (either `question`, `course-element`, or `core-element`).
```

Comment:
Is there a reason this distinction is important and these can't just all be executed the same way?

Reply:
Yes - the `type` determines a) where to resolve `directory` and b) whether or not to include `serverFilesCourse` on the Python interpreter's path.